{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example BIU Usage\n",
    "\n",
    "BIU is a toolkit I made to gather various datasets and tools that I regularly use.\n",
    "This way, I never need to worry about managing the data in files on my computer, I can simply use this package which has wrapper functions for common queries I perform on the datasets.\n",
    "\n",
    "Currently, it allows me to \n",
    "\n",
    " * Download a number of datasets on the fly - when they are needed (and sub-components of these datasets)\n",
    " * Dynamically load datasets when they are needed (They do not consume memory until a query is performed upon them\n",
    " * Query these datasets\n",
    " * Handle various data formats:\n",
    "   * FASTA\n",
    "   * GFF3\n",
    "   * GAF (GO annotation file)\n",
    "   * SQLite databases\n",
    "   * VCF (using the pyvcf package)\n",
    " * Map IDs between various indexing methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import biu as biu\n",
    "import matplotlib.pylab as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Change some default settings\n",
    "We can change the default location for all data storage, and turn off debug messages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We set the default data directory to be: '/exports/molepi/tgehrmann/data/'\n",
      "Turned OFF debug messages\n",
      "Turned ON debug messages\n"
     ]
    }
   ],
   "source": [
    "where = '/exports/molepi/tgehrmann/data/'\n",
    "biu.config.settings.setWhere(where)\n",
    "print(\"We set the default data directory to be: '%s'\" % biu.config.settings.getWhere())\n",
    "\n",
    "biu.config.settings.setDebugState(False)\n",
    "print(\"Turned OFF debug messages\")\n",
    "biu.config.settings.setDebugState(True)\n",
    "print(\"Turned ON debug messages\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## List the available datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "biu.db.list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Open a genome object and load the GFF file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the genome, and get the GFF file and parse it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hg = biu.db.Genome(\"Ensembl_91\")\n",
    "print(hg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqid = list(hg.aa.entries.keys())[1000]\n",
    "print('>%s\\n%s' % (seqid, hg.aa[seqid]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(hg.gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nTranscriptsPerGene = []\n",
    "nExonsPerTranscript = []\n",
    "for gene in hg.gff.topLevel['gene']:\n",
    "    transcripts = [ cid for (i, cid) in hg.gff.index[gene][1] ]\n",
    "    nTranscriptsPerGene.append(len(transcripts))\n",
    "    for trans in transcripts:\n",
    "        nExonsPerTranscript.append(len(hg.gff.index[trans][1]))\n",
    "    #efor\n",
    "#efor\n",
    "\n",
    "fig, axes = plt.subplots(figsize=(12,4), ncols=2, nrows=1)\n",
    "axes = axes.flatten()\n",
    "axes[0].hist(nTranscriptsPerGene, bins=20)\n",
    "axes[0].set_xlabel(\"Number of transcripts\")\n",
    "axes[0].set_ylabel(\"Number of genes\")\n",
    "axes[1].hist(nExonsPerTranscript, bins=20, log=True)\n",
    "axes[1].set_xlabel(\"Number of exons\")\n",
    "axes[1].set_ylabel(\"Number of genes\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Access the Uniprot database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot = biu.db.UniProt(\"human\")\n",
    "print(uniprot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for result in uniprot.getProteinDomains('P42345'):\n",
    "    print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Access the ClinVar database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cv = biu.db.ClinVar(\"GRCh37\")\n",
    "print(cv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alts = { n : 0 for n in 'ACGT'}\n",
    "for record in cv.queryVCF(1, 949422, 1049422):\n",
    "    for alt in record.ALT:\n",
    "        alt = alt.sequence\n",
    "        if alt in alts:\n",
    "            alts[alt] += 1\n",
    "\n",
    "cImpact = {}\n",
    "for record in cv.querySummary(1, 949422, 1049422):\n",
    "    if record.clinicalsignificance not in cImpact:\n",
    "        cImpact[record.clinicalsignificance] = 0\n",
    "    cImpact[record.clinicalsignificance] += 1\n",
    "\n",
    "fig, axes = plt.subplots(figsize=(12,4), ncols=2, nrows=1)\n",
    "axes = axes.flatten()\n",
    "\n",
    "nbars = axes[0].bar([1,2,3,4], alts.values(), tick_label=list(alts.keys()))\n",
    "\n",
    "nbars = axes[1].bar([ x + 1 for x in range(len(cImpact.keys())) ], cImpact.values(), tick_label=list(cImpact.keys()))\n",
    "plt.xticks(rotation=90)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Access the CADD database\n",
    "If you have pre-existing files elsewhere, you can tell the system where they are exactly with the \"localCopy\" argument, and it will make a symbolic link to our local copy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cadd = biu.db.CADD(localCopy = {\"tsv\" : \"/exports/molepi/tgehrmann/GAVIN-reimp/CADD/cadd.tsv.bgz\", \n",
    "                                \"tsv_tbi\" : \"/exports/molepi/tgehrmann/GAVIN-reimp/CADD/cadd.tsv.bgz.tbi\"})\n",
    "print(cadd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist([ float(p) for p in cadd.query(1, 0, 1000000).values() ], bins=200)\n",
    "plt.xlabel(\"CADD score\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Access the GnomAD database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad = biu.db.Gnomad(localCopy = { \"vcf\" : \"/exports/molepi/tgehrmann/GAVIN-reimp/gnomAD/gnomad.vcf.bgz\",\n",
    "                                     \"vcf_tbi\" : \"/exports/molepi/tgehrmann/GAVIN-reimp/gnomAD/gnomad.vcf.bgz.tbi\"})\n",
    "print(gnomad)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How to query from VCF files.\n",
    "There are different options to filter queries:\n",
    " * filters : Filter Variants based on a list of filters\n",
    " * gtFilters : Filter genotype calls based on a list of filters\n",
    " * types : Filter variants based on variant types\n",
    " * subTypes : Filter variants based on more specific variant types\n",
    " * sampleFilters : Filter sample calls based on a list of sample names. IMPORTANT: THIS WILL SELECT THOSE IN THE LIST, NOT FILTER THEM OUT!!\n",
    " \n",
    "These options can be used in any VCF query structure (e.g. clinVar, GnomAD, COSMIC, LLS, BBMRI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gnomad' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-4527bde8362a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mtypes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m \u001b[0;34m\"snp\"\u001b[0m \u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Without snp filter:\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgnomad\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m324719\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m324720\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfilters\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfilters\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"With snp filter:\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgnomad\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m324719\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m324720\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfilters\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfilters\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtypes\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtypes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"With RF filter:\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgnomad\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m324719\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m324720\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfilters\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfilters\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'RF'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtypes\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtypes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'gnomad' is not defined"
     ]
    }
   ],
   "source": [
    "filters = [ \"AMBIGUOUS\",\"VQLOW\",\"NVLOC\",\"CALLRATE\",\"MULTI\",\"RECMULTI\"]\n",
    "types = [ \"snp\" ]\n",
    "\n",
    "print(\"Without snp filter:\", len(list(gnomad.query(1, 324719, 324720, filters=filters))))\n",
    "print(\"With snp filter:\", len(list(gnomad.query(1, 324719, 324720, filters=filters, types=types))))\n",
    "print(\"With RF filter:\", len(list(gnomad.query(1, 324719, 324720, filters=filters + ['RF'], types=types))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D: Could not make Symbolic link for 'phen'. Rewriting internal location.\n",
      "D: Initializing the TSVResourceManager object NOW\n",
      "D: Initializing the VCFResourceManager object NOW\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>RR</th>\n",
       "      <th>R</th>\n",
       "      <th>RA</th>\n",
       "      <th>A</th>\n",
       "      <th>AA</th>\n",
       "      <th>U</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-26883511-A-C</td>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id   RR  R  RA  A  AA  U\n",
       "0  1-26883511-A-C  146  0  62  0  10  0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lls = biu.db.LLS(localCopy={\"phen\" : \"/home/tgehrmann/repos/VAR/phen218.txt\"})\n",
    "\n",
    "# Define the filters we want to use for the Variant calls:\n",
    "varFilters = [ \"AMBIGUOUS\",\"VQLOW\",\"NVLOC\",\"CALLRATE\",\"MULTI\",\"RECMULTI\"]\n",
    "\n",
    "# We are only interested in SNPs\n",
    "varTypes = [ \"snp\" ]\n",
    "\n",
    "# Within the LLS, we are only interested in 218 people, 4 of them we want to exclude (REMINDER: ask Erik why...)\n",
    "varSampleFilters = lls.phenotypes[\"cgID\"].apply(lambda x: x + '_240_37-ASM').values\n",
    "\n",
    "V = lls.query(1, 26883510, 26883511, filters=varFilters, types=varTypes, sampleFilters=varSampleFilters)\n",
    "S = lls.query(1, 26883510, 26883511, filters=varFilters, types=varTypes, sampleFilters=varSampleFilters, extract=\"summary\")\n",
    "\n",
    "biu.formats.VCF.summary(V)\n",
    "S\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TESTING\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D: Extract option: raw\n"
     ]
    }
   ],
   "source": [
    "V = lls.query(1, 26883510, 26883511, extract=\"raw\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alts = { n : 0 for n in 'ACGT'}\n",
    "for record in gnomad.queryVCF(1, 12590, 13000):\n",
    "    for alt in record.ALT:\n",
    "        alt = alt.sequence\n",
    "        if alt in alts:\n",
    "            alts[alt] += 1\n",
    "\n",
    "x = []; y = []\n",
    "for record in gnomad.queryCov(1, 12590, 13000, namedtuple=True):\n",
    "    x.append(int(record.pos))\n",
    "    y.append(float(record.mean))\n",
    "\n",
    "gene = \"gene:ENSG00000146648\"\n",
    "gEntry = hg.gff.getID(gene)\n",
    "genex = []; geney=[]\n",
    "for record in gnomad.queryCov(1, gEntry.start, gEntry.end, namedtuple=True):    \n",
    "    genex.append(int(record.pos))\n",
    "    geney.append(float(record.mean))\n",
    "\n",
    "fig, axes = plt.subplots(figsize=(12,4), ncols=3, nrows=1)\n",
    "axes = axes.flatten()\n",
    "\n",
    "nbars = axes[0].bar([1,2,3,4], alts.values(), tick_label=list(alts.keys()))\n",
    "axes[1].plot(x,y)\n",
    "axes[2].plot(genex,geney)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GTeX access\n",
    "Because GTeX is behind this ugly google login, you have to provide the files yourself.\n",
    "This can be done by specifying the exact location of the data.\n",
    "While the other datasets place the files deeper than the 'where' that you specify, GTeX will look EXACTLY for the following files:\n",
    " * `'where'/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.gct.gz`\n",
    " * `'where'/GTEx_Analysis_2016-01-15_v7_RSEMv1.2.22_transcript_tpm.txt.gz`\n",
    " * `'where'/GTEx_v7_Annotations_SampleAttributesDS.txt`\n",
    " * `'where'/GTEx_v7_Annotations_SubjectPhenotypesDS.txt`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "where = '/exports/molepi/tgehrmann/data/'\n",
    "import biu as biu\n",
    "biu.config.settings.setWhere(where)\n",
    "gtex = biu.db.GTeX(version=\"v7\",\n",
    "                   where=\"/exports/molepi/tgehrmann/data/gtex\")\n",
    "print(gtex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(gtex.getPersonIDSamples(gtex.getPersonIDs()[0]))\n",
    "%time gtex.getGeneExpr(gtex.getPersonIDSamples(gtex.getPersonIDs()[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pTissues = {}\n",
    "for i, row in gtex.sAttr.iterrows():\n",
    "    if row[\"SMAFRZE\"] != \"RNASEQ\":\n",
    "        continue\n",
    "    #fi\n",
    "    personID = row[\"SAMPID\"].split('-')[1]\n",
    "    sampleType = row[\"SMTSD\"]\n",
    "    if personID not in pTissues:\n",
    "        pTissues[personID] = []\n",
    "    pTissues[personID].append(sampleType)\n",
    "\n",
    "indivTissues = sorted(list(set(gtex.sAttr[\"SMTSD\"])))\n",
    "pairwiseTissueCounts = {}\n",
    "for personID in pTissues:\n",
    "    tissues = list(set(pTissues[personID]))\n",
    "    for i, samplei in enumerate(tissues[:-1]):\n",
    "        for j, samplej in enumerate(tissues[i+1:]):\n",
    "            key = (samplei, samplej)\n",
    "            if key not in pairwiseTissueCounts:\n",
    "                pairwiseTissueCounts[key] = 0\n",
    "            pairwiseTissueCounts[(samplei, samplej)] += 1\n",
    "\n",
    "indivTissuesMap = { t: i for (i,t) in enumerate(list(indivTissues)) }\n",
    "\n",
    "C = np.zeros([len(indivTissues), len(indivTissues)])\n",
    "for (t1,t2) in pairwiseTissueCounts:\n",
    "    C[indivTissuesMap[t1], indivTissuesMap[t2]] = int(pairwiseTissueCounts.get((t1,t2),0) +\n",
    "                                                  pairwiseTissueCounts.get((t2,t1),0))\n",
    "    C[indivTissuesMap[t2], indivTissuesMap[t1]] = int(pairwiseTissueCounts.get((t1,t2),0) +\n",
    "                                                  pairwiseTissueCounts.get((t2,t1),0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(40,40))\n",
    "sns.heatmap(C, ax = ax, xticklabels=indivTissues, yticklabels=indivTissues, annot=True, fmt='.0f')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interestTissues = [ 'Adipose - Subcutaneous', 'Adipose - Visceral (Omentum)', 'Muscle - Skeletal', \"Whole Blood\" ]\n",
    "interestTissuesIndex = [ indivTissuesMap[t] for t in interestTissues ]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10,10))\n",
    "sns.heatmap(C[interestTissuesIndex,:][:,interestTissuesIndex], ax = ax, xticklabels=interestTissues, yticklabels=interestTissues, annot=True, fmt='.1f')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reactome access\n",
    "Doesn't really work yet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reactome = biu.db.Reactome(where=where + '/reactome')\n",
    "#print(reactome)\n",
    "#for r in reactome.getPathway(\"R-HSA-1236973\"):\n",
    "#    print(r)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cosmic access"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cosmic = biu.db.Cosmic(\"t.gehrmann@lumc.nl\", \"cab847t0\")\n",
    "print(cosmic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for r in cosmic.vcfCoding.query(1, 1, 69270):\n",
    "    print(r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for r in cosmic.vcfNonCoding.query(1, 1, 69270):\n",
    "    print(r)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Do gene mapping with pickled maps\n",
    "Faster operations, but slow initialization + more memory usage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import biu\n",
    "hm = biu.maps.Human(where=\"/exports/molepi/tgehrmann/data/\")\n",
    "print(hm)\n",
    "\n",
    "def exampleMapping(GMO):\n",
    "    # GMO : Gene Mapping Object\n",
    "    symbol = \"MTOR\"\n",
    "    geneid = GMO.getSymbolGeneID(symbol)[0]\n",
    "    print(\"%s -> %s\" % (symbol, geneid))\n",
    "    symbol = GMO.getGeneIDSymbol(geneid)[0]\n",
    "    print(\"%s -> %s\" % (geneid, symbol))\n",
    "    ensembl = GMO.getSymbolEnsembl(symbol)[0]\n",
    "    print(\"%s -> %s\" % (symbol, ensembl))\n",
    "    symbol = GMO.getEnsemblSymbol(ensembl)[0]\n",
    "    print(\"%s -> %s\" % (ensembl, symbol))\n",
    "#edef\n",
    "\n",
    "def exampleMappingSilent(GMO):\n",
    "    # GMO : Gene Mapping Object\n",
    "    symbol = \"MTOR\"\n",
    "    geneid = GMO.getSymbolGeneID(symbol)[0]\n",
    "    symbol = GMO.getGeneIDSymbol(geneid)[0]\n",
    "    ensembl = GMO.getSymbolEnsembl(symbol)[0]\n",
    "    symbol = GMO.getEnsemblSymbol(ensembl)[0]\n",
    "#edef"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleMapping(hm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping with SQLite instead of pickled Maps\n",
    "Fast initialization, but slower operations.\n",
    "Because of the high speed initialization, we can perform queries on a larger number of structures, including the gene2refseq index, and the uniprotmap, which is prohibitively large for the map."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "where = '/exports/molepi/tgehrmann/data/'\n",
    "import biu as biu\n",
    "biu.config.settings.setWhere(where)\n",
    "print(biu.config.settings.getWhere())\n",
    "hms = biu.maps.HumanS()\n",
    "print(hms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleMapping(hms)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare Map vs SQLite speeds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Map Lookup\")\n",
    "%timeit exampleMappingSilent(hm)\n",
    "print(\"SQLite lookup\")\n",
    "%timeit exampleMappingSilent(hms)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HAGR access\n",
    "Unlike others, HAGR is downloaded already entirely when the class is initiated (because they are in ZIP files, and I don't have a nice solution for this yet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import biu as biu\n",
    "hagr = biu.db.HAGR(where = '/exports/molepi/tgehrmann/data/')\n",
    "print(hagr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hagr.human_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Access GO annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "where = '/exports/molepi/tgehrmann/data/'\n",
    "import biu as biu\n",
    "biu.config.settings.setWhere(where)\n",
    "print(biu.config.settings.getWhere())\n",
    "go = biu.db.GO()\n",
    "print(go)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Number of genes annotated with GO:0002250: %d\" % len(go.getAnnotated(\"GO:0002250\")))\n",
    "\n",
    "print(\"Number of annotations for P78540: %d\" % len(go.getAnnots(\"P78540\")))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Access KEGG annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "where = '/exports/molepi/tgehrmann/data/'\n",
    "import biu as biu\n",
    "biu.config.settings.setWhere(where)\n",
    "print(biu.config.settings.getWhere())\n",
    "kegg = biu.db.KEGG()\n",
    "hms = biu.maps.HumanS()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(kegg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Number of pathways MTOR is in: %d\" % len(kegg.getGenePathways(hms.getSymbolGeneID(\"MTOR\")[0])))\n",
    "\n",
    "print(\"Number of genes in path:hsa05230: %d\" % len(kegg.getPathwayGenes(\"path:hsa05230\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(kegg.getPathwayInfo(\"hsa05230\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(figsize=(12,4), ncols=2, nrows=1)\n",
    "axes = axes.flatten()\n",
    "\n",
    "# How many genes are there per kegg pathway?\n",
    "genesPerPathway = [ len(kegg.getPathwayGenes(p)) for p in kegg.getPathways() ]\n",
    "pathwaysPerGene = [ len(kegg.getGenePathways(g)) for g in kegg.getGenes() ]\n",
    "\n",
    "axes[0].hist(genesPerPathway, bins=50)\n",
    "axes[0].set_xlabel(\"Number of genes per pathway\")\n",
    "axes[1].hist(pathwaysPerGene, bins=50)\n",
    "axes[1].set_xlabel(\"Number of pathways per gene\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Access LLS data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lls = biu.db.LLS()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for r in lls.queryRegions([ (\"1\", 100, 100000), (\"1\", 100, 100000)]):\n",
    "    print(r)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Access BBMRI data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bbmri = biu.db.BBMRI()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Store some variables persistently in a SQLite database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pDict = biu.formats.SQLDict(\"test\")\n",
    "print(pDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pDict[\"test\"] = { 5: \"hello\", \"aha\" : [ 1, 4, \"345\"]}\n",
    "print(\"test -> \", pDict[\"test\"])\n",
    "print(\"yest -> \", pDict[\"yest\"])\n",
    "\n",
    "for x in pDict:\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Access miRmine database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import biu\n",
    "where = '/exports/molepi/tgehrmann/data/'\n",
    "biu.config.settings.setWhere(where)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mir = biu.db.MiRmine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(mir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mir.getExpr([\"DRX003170\", \"DRX003171\", \"DRX017209\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set(mir._info[\"Tissue\"][mir._info[\"Tissue\"].apply(lambda x: not(pd.isnull(x)))].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Use pipelines"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use the VEP pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import biu as biu\n",
    "where = '/exports/molepi/tgehrmann/data/'\n",
    "biu.config.settings.setWhere(where)\n",
    "\n",
    "lls = biu.db.LLS()\n",
    "varList = list([ r for r in lls.query(1, 10483, 10495)])\n",
    "vep = biu.pipelines.VEP(varList)\n",
    "vep.getAnnotations()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use the LiftOver Pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import biu as biu\n",
    "where = '/exports/molepi/tgehrmann/data/'\n",
    "biu.config.settings.setWhere(where)\n",
    "\n",
    "lls = biu.db.LLS()\n",
    "varList = [ (r.CHROM, r.POS-1, r.POS) for r in lls.query(5, 42423775, 42426000, filt=['VQLOW'])]\n",
    "varLift = biu.pipelines.LiftOver(varList)\n",
    "varLift.getLiftOver()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "varLift.getLiftOver().values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(lls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D: Same symbolic link already exists for 'vcf'\n",
      "D: Same symbolic link already exists for 'vcf_tbi'\n",
      "D: Initializing the VCFResourceManager object NOW\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n",
      "0 1 0 1 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>RR</th>\n",
       "      <th>R</th>\n",
       "      <th>RA</th>\n",
       "      <th>A</th>\n",
       "      <th>AA</th>\n",
       "      <th>U</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700109-C-T</td>\n",
       "      <td>20000</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700114-C-A</td>\n",
       "      <td>20966</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700116-C-T</td>\n",
       "      <td>21353</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700117-C-G</td>\n",
       "      <td>8013</td>\n",
       "      <td>0</td>\n",
       "      <td>10003</td>\n",
       "      <td>0</td>\n",
       "      <td>3793</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700119-G-A</td>\n",
       "      <td>22756</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700128-G-A</td>\n",
       "      <td>26373</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700135-G-T</td>\n",
       "      <td>33561</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700140-G-A</td>\n",
       "      <td>36546</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700145-G-C</td>\n",
       "      <td>38049</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700146-G-T</td>\n",
       "      <td>38314</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700147-G-A</td>\n",
       "      <td>37889</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700148-C-T</td>\n",
       "      <td>37622</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700149-G-A</td>\n",
       "      <td>37603</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700151-C-T</td>\n",
       "      <td>39472</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700153-C-T</td>\n",
       "      <td>39961</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700157-C-A</td>\n",
       "      <td>41737</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700163-G-A</td>\n",
       "      <td>46765</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700173-AG-A</td>\n",
       "      <td>49706</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700184-C-T</td>\n",
       "      <td>53835</td>\n",
       "      <td>0</td>\n",
       "      <td>73</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700185-G-A</td>\n",
       "      <td>54291</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700186-C-T</td>\n",
       "      <td>55016</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700188-C-T</td>\n",
       "      <td>56620</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700189-G-A</td>\n",
       "      <td>57145</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700191-C-T</td>\n",
       "      <td>57740</td>\n",
       "      <td>0</td>\n",
       "      <td>199</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700193-A-C</td>\n",
       "      <td>59125</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700198-G-A</td>\n",
       "      <td>64846</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700199-T-A</td>\n",
       "      <td>65201</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700201-C-T</td>\n",
       "      <td>65322</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700206-A-C</td>\n",
       "      <td>66516</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2700207-C-T</td>\n",
       "      <td>66458</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706106-C-T</td>\n",
       "      <td>48525</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706109-CCGGCCCTGCGCTCTGGA-C</td>\n",
       "      <td>48620</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706110-C-T</td>\n",
       "      <td>47218</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706111-G-A</td>\n",
       "      <td>47319</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706119-G-A</td>\n",
       "      <td>45507</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706123-T-C</td>\n",
       "      <td>44602</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706127-C-A</td>\n",
       "      <td>42884</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706128-G-C</td>\n",
       "      <td>41757</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706135-C-G</td>\n",
       "      <td>41212</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706138-G-C</td>\n",
       "      <td>40591</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706140-G-A</td>\n",
       "      <td>40413</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706146-C-T</td>\n",
       "      <td>37868</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706147-G-A</td>\n",
       "      <td>32498</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706149-T-C</td>\n",
       "      <td>31292</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706155-A-G</td>\n",
       "      <td>25304</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706156-T-A</td>\n",
       "      <td>24288</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706162-G-C</td>\n",
       "      <td>17192</td>\n",
       "      <td>0</td>\n",
       "      <td>182</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706171-C-T</td>\n",
       "      <td>11572</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706173-C-A</td>\n",
       "      <td>11254</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706174-G-A</td>\n",
       "      <td>11258</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706177-G-A</td>\n",
       "      <td>8917</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706179-A-G</td>\n",
       "      <td>8986</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706181-C-T</td>\n",
       "      <td>7912</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706183-C-G</td>\n",
       "      <td>7095</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706186-G-C</td>\n",
       "      <td>6600</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706189-G-T</td>\n",
       "      <td>5895</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706192-GC-G</td>\n",
       "      <td>4917</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706193-C-T</td>\n",
       "      <td>4974</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706195-C-A</td>\n",
       "      <td>4526</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2706203-T-A</td>\n",
       "      <td>3794</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>298 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                id     RR  R     RA  A    AA  U\n",
       "0                    1-2700109-C-T  20000  0      2  0     0  0\n",
       "0                    1-2700114-C-A  20966  0      7  0     0  0\n",
       "0                    1-2700116-C-T  21353  0      1  0     0  0\n",
       "0                    1-2700117-C-G   8013  0  10003  0  3793  0\n",
       "0                    1-2700119-G-A  22756  0      0  0     0  0\n",
       "0                    1-2700128-G-A  26373  0      1  0     0  0\n",
       "0                    1-2700135-G-T  33561  0      0  0     0  0\n",
       "0                    1-2700140-G-A  36546  0      1  0     0  0\n",
       "0                    1-2700145-G-C  38049  0      1  0     0  0\n",
       "0                    1-2700146-G-T  38314  0      1  0     0  0\n",
       "0                    1-2700147-G-A  37889  0      1  0     0  0\n",
       "0                    1-2700148-C-T  37622  0      6  0     0  0\n",
       "0                    1-2700149-G-A  37603  0      3  0     0  0\n",
       "0                    1-2700151-C-T  39472  0      0  0     0  0\n",
       "0                    1-2700153-C-T  39961  0      1  0     0  0\n",
       "0                    1-2700157-C-A  41737  0      2  0     0  0\n",
       "0                    1-2700163-G-A  46765  0      0  0     0  0\n",
       "0                   1-2700173-AG-A  49706  0      1  0     0  0\n",
       "0                    1-2700184-C-T  53835  0     73  0     3  0\n",
       "0                    1-2700185-G-A  54291  0      0  0     0  0\n",
       "0                    1-2700186-C-T  55016  0      0  0     0  0\n",
       "0                    1-2700188-C-T  56620  0      0  0     0  0\n",
       "0                    1-2700189-G-A  57145  0      8  0     0  0\n",
       "0                    1-2700191-C-T  57740  0    199  0     1  0\n",
       "0                    1-2700193-A-C  59125  0      2  0     0  0\n",
       "0                    1-2700198-G-A  64846  0      1  0     0  0\n",
       "0                    1-2700199-T-A  65201  0     19  0     0  0\n",
       "0                    1-2700201-C-T  65322  0      1  0     0  0\n",
       "0                    1-2700206-A-C  66516  0      0  0     0  0\n",
       "0                    1-2700207-C-T  66458  0      4  0     0  0\n",
       "..                             ...    ... ..    ... ..   ... ..\n",
       "0                    1-2706106-C-T  48525  0      1  0     0  0\n",
       "0   1-2706109-CCGGCCCTGCGCTCTGGA-C  48620  0      1  0     0  0\n",
       "0                    1-2706110-C-T  47218  0      3  0     0  0\n",
       "0                    1-2706111-G-A  47319  0      1  0     0  0\n",
       "0                    1-2706119-G-A  45507  0      3  0     0  0\n",
       "0                    1-2706123-T-C  44602  0      2  0     0  0\n",
       "0                    1-2706127-C-A  42884  0      0  0     0  0\n",
       "0                    1-2706128-G-C  41757  0      1  0     0  0\n",
       "0                    1-2706135-C-G  41212  0      3  0     0  0\n",
       "0                    1-2706138-G-C  40591  0      5  0     0  0\n",
       "0                    1-2706140-G-A  40413  0     14  0     0  0\n",
       "0                    1-2706146-C-T  37868  0      1  0     0  0\n",
       "0                    1-2706147-G-A  32498  0      0  0     0  0\n",
       "0                    1-2706149-T-C  31292  0      0  0     0  0\n",
       "0                    1-2706155-A-G  25304  0      1  0     0  0\n",
       "0                    1-2706156-T-A  24288  0     72  0     1  0\n",
       "0                    1-2706162-G-C  17192  0    182  0     1  0\n",
       "0                    1-2706171-C-T  11572  0      2  0     0  0\n",
       "0                    1-2706173-C-A  11254  0      0  0     0  0\n",
       "0                    1-2706174-G-A  11258  0      0  0     0  0\n",
       "0                    1-2706177-G-A   8917  0      0  0     0  0\n",
       "0                    1-2706179-A-G   8986  0      0  0     0  0\n",
       "0                    1-2706181-C-T   7912  0      0  0     0  0\n",
       "0                    1-2706183-C-G   7095  0      3  0     0  0\n",
       "0                    1-2706186-G-C   6600  0      0  0     0  0\n",
       "0                    1-2706189-G-T   5895  0      1  0     0  0\n",
       "0                   1-2706192-GC-G   4917  0      0  0     0  0\n",
       "0                    1-2706193-C-T   4974  0      0  0     0  0\n",
       "0                    1-2706195-C-A   4526  0      0  0     0  0\n",
       "0                    1-2706203-T-A   3794  0      0  0     0  0\n",
       "\n",
       "[298 rows x 7 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import biu as biu\n",
    "import matplotlib.pylab as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "gnomad = biu.db.Gnomad(localCopy = { \"vcf\" : \"/exports/molepi/tgehrmann/GAVIN-reimp/gnomAD/gnomad.vcf.bgz\",\n",
    "                                     \"vcf_tbi\" : \"/exports/molepi/tgehrmann/GAVIN-reimp/gnomAD/gnomad.vcf.bgz.tbi\"})\n",
    "\n",
    "filters = [ \"AMBIGUOUS\",\"VQLOW\",\"NVLOC\",\"CALLRATE\",\"MULTI\",\"RECMULTI\"]\n",
    "types = [ \"snp\" ]\n",
    "\n",
    "gnomad.query('Y', 2655002, 2655003)\n",
    "gnomad.query('Y', 2655024, 2655025)\n",
    "v = gnomad.query('Y', 2709618, 2709619)\n",
    "#gnomad.summary(v, [1])\n",
    "\n",
    "gnomad.query('1', 2655002, 2709619, extract=\"summary\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
