{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Accessing the LLS dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import biu"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initializing and examining the structure\n",
    "The phenotypes data file is not located in a reasonable location. We must provide our own local copy of it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LLS object\n",
      " Where: /exports/molepi/LLSSEQ\n",
      " Version: current\n",
      " Objects:\n",
      "  * [ ] phenotypes\n",
      "  * [ ] vcf[1]\n",
      "  * [ ] vcf[2]\n",
      "  * [ ] vcf[3]\n",
      "  * [ ] vcf[4]\n",
      "  * [ ] vcf[5]\n",
      "  * [ ] vcf[6]\n",
      "  * [ ] vcf[7]\n",
      "  * [ ] vcf[8]\n",
      "  * [ ] vcf[9]\n",
      "  * [ ] vcf[10]\n",
      "  * [ ] vcf[11]\n",
      "  * [ ] vcf[12]\n",
      "  * [ ] vcf[13]\n",
      "  * [ ] vcf[14]\n",
      "  * [ ] vcf[15]\n",
      "  * [ ] vcf[16]\n",
      "  * [ ] vcf[17]\n",
      "  * [ ] vcf[18]\n",
      "  * [ ] vcf[19]\n",
      "  * [ ] vcf[20]\n",
      "  * [ ] vcf[21]\n",
      "  * [ ] vcf[22]\n",
      "  * [ ] vcf[M]\n",
      "  * [ ] vcf[X]\n",
      "  * [ ] vcf[Y]\n",
      " Files:\n",
      "  * [X] vcf_1 : /exports/molepi/LLSSEQ/tbx/merged.chr1.vcf.bgz\n",
      "  * [X] vcf_1_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr1.vcf.bgz.tbi\n",
      "  * [X] vcf_2 : /exports/molepi/LLSSEQ/tbx/merged.chr2.vcf.bgz\n",
      "  * [X] vcf_2_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr2.vcf.bgz.tbi\n",
      "  * [X] vcf_3 : /exports/molepi/LLSSEQ/tbx/merged.chr3.vcf.bgz\n",
      "  * [X] vcf_3_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr3.vcf.bgz.tbi\n",
      "  * [X] vcf_4 : /exports/molepi/LLSSEQ/tbx/merged.chr4.vcf.bgz\n",
      "  * [X] vcf_4_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr4.vcf.bgz.tbi\n",
      "  * [X] vcf_5 : /exports/molepi/LLSSEQ/tbx/merged.chr5.vcf.bgz\n",
      "  * [X] vcf_5_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr5.vcf.bgz.tbi\n",
      "  * [X] vcf_6 : /exports/molepi/LLSSEQ/tbx/merged.chr6.vcf.bgz\n",
      "  * [X] vcf_6_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr6.vcf.bgz.tbi\n",
      "  * [X] vcf_7 : /exports/molepi/LLSSEQ/tbx/merged.chr7.vcf.bgz\n",
      "  * [X] vcf_7_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr7.vcf.bgz.tbi\n",
      "  * [X] vcf_8 : /exports/molepi/LLSSEQ/tbx/merged.chr8.vcf.bgz\n",
      "  * [X] vcf_8_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr8.vcf.bgz.tbi\n",
      "  * [X] vcf_9 : /exports/molepi/LLSSEQ/tbx/merged.chr9.vcf.bgz\n",
      "  * [X] vcf_9_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr9.vcf.bgz.tbi\n",
      "  * [X] vcf_10 : /exports/molepi/LLSSEQ/tbx/merged.chr10.vcf.bgz\n",
      "  * [X] vcf_10_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr10.vcf.bgz.tbi\n",
      "  * [X] vcf_11 : /exports/molepi/LLSSEQ/tbx/merged.chr11.vcf.bgz\n",
      "  * [X] vcf_11_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr11.vcf.bgz.tbi\n",
      "  * [X] vcf_12 : /exports/molepi/LLSSEQ/tbx/merged.chr12.vcf.bgz\n",
      "  * [X] vcf_12_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr12.vcf.bgz.tbi\n",
      "  * [X] vcf_13 : /exports/molepi/LLSSEQ/tbx/merged.chr13.vcf.bgz\n",
      "  * [X] vcf_13_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr13.vcf.bgz.tbi\n",
      "  * [X] vcf_14 : /exports/molepi/LLSSEQ/tbx/merged.chr14.vcf.bgz\n",
      "  * [X] vcf_14_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr14.vcf.bgz.tbi\n",
      "  * [X] vcf_15 : /exports/molepi/LLSSEQ/tbx/merged.chr15.vcf.bgz\n",
      "  * [X] vcf_15_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr15.vcf.bgz.tbi\n",
      "  * [X] vcf_16 : /exports/molepi/LLSSEQ/tbx/merged.chr16.vcf.bgz\n",
      "  * [X] vcf_16_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr16.vcf.bgz.tbi\n",
      "  * [X] vcf_17 : /exports/molepi/LLSSEQ/tbx/merged.chr17.vcf.bgz\n",
      "  * [X] vcf_17_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr17.vcf.bgz.tbi\n",
      "  * [X] vcf_18 : /exports/molepi/LLSSEQ/tbx/merged.chr18.vcf.bgz\n",
      "  * [X] vcf_18_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr18.vcf.bgz.tbi\n",
      "  * [X] vcf_19 : /exports/molepi/LLSSEQ/tbx/merged.chr19.vcf.bgz\n",
      "  * [X] vcf_19_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr19.vcf.bgz.tbi\n",
      "  * [X] vcf_20 : /exports/molepi/LLSSEQ/tbx/merged.chr20.vcf.bgz\n",
      "  * [X] vcf_20_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr20.vcf.bgz.tbi\n",
      "  * [X] vcf_21 : /exports/molepi/LLSSEQ/tbx/merged.chr21.vcf.bgz\n",
      "  * [X] vcf_21_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr21.vcf.bgz.tbi\n",
      "  * [X] vcf_22 : /exports/molepi/LLSSEQ/tbx/merged.chr22.vcf.bgz\n",
      "  * [X] vcf_22_tbi : /exports/molepi/LLSSEQ/tbx/merged.chr22.vcf.bgz.tbi\n",
      "  * [X] vcf_M : /exports/molepi/LLSSEQ/tbx/merged.chrM.vcf.bgz\n",
      "  * [X] vcf_M_tbi : /exports/molepi/LLSSEQ/tbx/merged.chrM.vcf.bgz.tbi\n",
      "  * [X] vcf_X : /exports/molepi/LLSSEQ/tbx/merged.chrX.vcf.bgz\n",
      "  * [X] vcf_X_tbi : /exports/molepi/LLSSEQ/tbx/merged.chrX.vcf.bgz.tbi\n",
      "  * [X] vcf_Y : /exports/molepi/LLSSEQ/tbx/merged.chrY.vcf.bgz\n",
      "  * [X] vcf_Y_tbi : /exports/molepi/LLSSEQ/tbx/merged.chrY.vcf.bgz.tbi\n",
      "  * [X] phen : /home/tgehrmann/repos/VAR/phen218.txt\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D: Could not make Symbolic link for 'phen'. Rewriting internal location.\n"
     ]
    }
   ],
   "source": [
    "lls = biu.db.LLS(localCopy={\"phen\" : \"/home/tgehrmann/repos/VAR/phen218.txt\"})\n",
    "print(lls)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Querying the structure\n",
    "\n",
    "With the exception of the list of participants, the data held in LLS are tabix indexed VCF files. Thus, we can make use of the `query` and `queryRegions` functionalities of the VCF structures.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Examining the list of participants\n",
    "The `phenotypes` object of the lls structure describes the list of participants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cgID</th>\n",
       "      <th>cgID2</th>\n",
       "      <th>LLnr</th>\n",
       "      <th>labnr</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GS00354-DNA_B01</td>\n",
       "      <td>GS000002814-ASM</td>\n",
       "      <td>005.2.01.0</td>\n",
       "      <td>6</td>\n",
       "      <td>100.974675</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>GS00354-DNA_F07</td>\n",
       "      <td>GS000002795-ASM</td>\n",
       "      <td>112.2.02.0</td>\n",
       "      <td>695</td>\n",
       "      <td>102.193018</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>GS00398-DNA_C02</td>\n",
       "      <td>GS000003087-ASM</td>\n",
       "      <td>429.2.01.0</td>\n",
       "      <td>2822</td>\n",
       "      <td>100.517454</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>GS00398-DNA_B01</td>\n",
       "      <td>GS000002935-ASM</td>\n",
       "      <td>441.2.02.0</td>\n",
       "      <td>2740</td>\n",
       "      <td>101.141684</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>GS00398-DNA_C03</td>\n",
       "      <td>GS000003081-ASM</td>\n",
       "      <td>468.2.02.0</td>\n",
       "      <td>2895</td>\n",
       "      <td>100.183436</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                cgID            cgID2        LLnr  labnr         age   sex\n",
       "1    GS00354-DNA_B01  GS000002814-ASM  005.2.01.0      6  100.974675  MALE\n",
       "52   GS00354-DNA_F07  GS000002795-ASM  112.2.02.0    695  102.193018  MALE\n",
       "188  GS00398-DNA_C02  GS000003087-ASM  429.2.01.0   2822  100.517454  MALE\n",
       "196  GS00398-DNA_B01  GS000002935-ASM  441.2.02.0   2740  101.141684  MALE\n",
       "211  GS00398-DNA_C03  GS000003081-ASM  468.2.02.0   2895  100.183436  MALE"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lls.phenotypes[(lls.phenotypes[\"sex\"] == \"MALE\") & (lls.phenotypes[\"age\"] > 100) ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Querying the variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record(CHROM=1, POS=1000760, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=1000797, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=1000854, REF=A, ALT=[C])\n",
      "Record(CHROM=1, POS=1000857, REF=G, ALT=[T])\n",
      "Record(CHROM=1, POS=1000894, REF=A, ALT=[T])\n",
      "Record(CHROM=1, POS=1000902, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=1000910, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=1000987, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=1000990, REF=T, ALT=[G])\n"
     ]
    }
   ],
   "source": [
    "for record in lls.query(1, 1000000, 1001000, types=['snp']):\n",
    "    print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record(CHROM=1, POS=1000760, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=1000797, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=1000854, REF=A, ALT=[C])\n",
      "Record(CHROM=1, POS=1000857, REF=G, ALT=[T])\n",
      "Record(CHROM=1, POS=1000894, REF=A, ALT=[T])\n",
      "Record(CHROM=1, POS=1000902, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=1000910, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=1000987, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=1000990, REF=T, ALT=[G])\n",
      "Record(CHROM=2, POS=1000024, REF=T, ALT=[C])\n",
      "Record(CHROM=2, POS=1000029, REF=G, ALT=[A])\n",
      "Record(CHROM=2, POS=1000127, REF=C, ALT=[T])\n",
      "Record(CHROM=2, POS=1000192, REF=A, ALT=[T])\n",
      "Record(CHROM=2, POS=1000283, REF=C, ALT=[T])\n",
      "Record(CHROM=2, POS=1000738, REF=A, ALT=[G])\n",
      "Record(CHROM=2, POS=1000918, REF=T, ALT=[A])\n"
     ]
    }
   ],
   "source": [
    "for record in lls.queryRegions([ (1, 1000000, 1001000), (2, 1000000, 1001000)], types=['snp']):\n",
    "    print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>RR</th>\n",
       "      <th>R</th>\n",
       "      <th>RA</th>\n",
       "      <th>A</th>\n",
       "      <th>AA</th>\n",
       "      <th>O</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-1000760-G-A</td>\n",
       "      <td>221</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1-1000797-C-T</td>\n",
       "      <td>218</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1-1000854-A-C</td>\n",
       "      <td>214</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1-1000857-G-T</td>\n",
       "      <td>212</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1-1000894-A-T</td>\n",
       "      <td>214</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1-1000902-G-A</td>\n",
       "      <td>221</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1-1000910-C-T</td>\n",
       "      <td>172</td>\n",
       "      <td>0</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1-1000987-T-C</td>\n",
       "      <td>221</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1-1000990-T-G</td>\n",
       "      <td>221</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2-1000024-T-C</td>\n",
       "      <td>212</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2-1000029-G-A</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>110</td>\n",
       "      <td>0</td>\n",
       "      <td>51</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2-1000127-C-T</td>\n",
       "      <td>172</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2-1000192-A-T</td>\n",
       "      <td>221</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2-1000283-C-T</td>\n",
       "      <td>165</td>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2-1000738-A-G</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2-1000918-T-A</td>\n",
       "      <td>220</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id   RR  R   RA  A  AA  O\n",
       "0   1-1000760-G-A  221  0    1  0   0  0\n",
       "1   1-1000797-C-T  218  0    1  0   0  3\n",
       "2   1-1000854-A-C  214  0    1  0   0  7\n",
       "3   1-1000857-G-T  212  0    1  0   0  9\n",
       "4   1-1000894-A-T  214  0    7  0   0  1\n",
       "5   1-1000902-G-A  221  0    1  0   0  0\n",
       "6   1-1000910-C-T  172  0   39  0  11  0\n",
       "7   1-1000987-T-C  221  0    1  0   0  0\n",
       "8   1-1000990-T-G  221  0    1  0   0  0\n",
       "9   2-1000024-T-C  212  0   10  0   0  0\n",
       "10  2-1000029-G-A   61  0  110  0  51  0\n",
       "11  2-1000127-C-T  172  0   48  0   2  0\n",
       "12  2-1000192-A-T  221  0    1  0   0  0\n",
       "13  2-1000283-C-T  165  0   55  0   2  0\n",
       "14  2-1000738-A-G  194  0   27  0   1  0\n",
       "15  2-1000918-T-A  220  0    2  0   0  0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lls.queryRegions([ (1, 1000000, 1001000), (2, 1000000, 1001000)], types=['snp'], extract=\"summary\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Querying the variants based on specific samples\n",
    "The sample names provided in the VCF files of BBMRI are not the same as in the phenotypes file.\n",
    "The names in the VCF have `'_240_37-ASM'` added at the end. To focus on only those, we must filter on them correctly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>RR</th>\n",
       "      <th>R</th>\n",
       "      <th>RA</th>\n",
       "      <th>A</th>\n",
       "      <th>AA</th>\n",
       "      <th>O</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-1000760-G-A</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1-1000797-C-T</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1-1000854-A-C</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1-1000857-G-T</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1-1000894-A-T</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1-1000902-G-A</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1-1000910-C-T</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1-1000987-T-C</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1-1000990-T-G</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2-1000024-T-C</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2-1000029-G-A</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2-1000127-C-T</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2-1000192-A-T</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2-1000283-C-T</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2-1000738-A-G</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2-1000918-T-A</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id  RR  R  RA  A  AA  O\n",
       "0   1-1000760-G-A   5  0   0  0   0  0\n",
       "1   1-1000797-C-T   5  0   0  0   0  0\n",
       "2   1-1000854-A-C   4  0   0  0   0  1\n",
       "3   1-1000857-G-T   4  0   0  0   0  1\n",
       "4   1-1000894-A-T   5  0   0  0   0  0\n",
       "5   1-1000902-G-A   5  0   0  0   0  0\n",
       "6   1-1000910-C-T   3  0   1  0   1  0\n",
       "7   1-1000987-T-C   5  0   0  0   0  0\n",
       "8   1-1000990-T-G   5  0   0  0   0  0\n",
       "9   2-1000024-T-C   5  0   0  0   0  0\n",
       "10  2-1000029-G-A   1  0   3  0   1  0\n",
       "11  2-1000127-C-T   3  0   2  0   0  0\n",
       "12  2-1000192-A-T   5  0   0  0   0  0\n",
       "13  2-1000283-C-T   4  0   1  0   0  0\n",
       "14  2-1000738-A-G   5  0   0  0   0  0\n",
       "15  2-1000918-T-A   5  0   0  0   0  0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "relSamples = lls.phenotypes[(lls.phenotypes[\"sex\"] == \"MALE\") & (lls.phenotypes[\"age\"] > 100) ]\n",
    "sampleFilters = relSamples[\"cgID\"].apply(lambda x: x + '_240_37-ASM').values\n",
    "sampleFilters\n",
    "\n",
    "lls.queryRegions([ (1, 1000000, 1001000), (2, 1000000, 1001000)], types=['snp'], \n",
    "                 extract=\"summary\", sampleFilters=sampleFilters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
