{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Accessing the GnomAD dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import biu\n",
    "\n",
    "# Set the default data location\n",
    "where = '/exports/molepi/tgehrmann/data/'\n",
    "biu.config.settings.setWhere(where)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gnomad object\n",
      " Version: GRCh37\n",
      " Objects:\n",
      "  * [ ] _vcf\n",
      "  * [ ] _cov_1\n",
      "  * [ ] _cov_2\n",
      "  * [ ] _cov_3\n",
      "  * [ ] _cov_4\n",
      "  * [ ] _cov_5\n",
      "  * [ ] _cov_6\n",
      "  * [ ] _cov_7\n",
      "  * [ ] _cov_8\n",
      "  * [ ] _cov_9\n",
      "  * [ ] _cov_10\n",
      "  * [ ] _cov_11\n",
      "  * [ ] _cov_12\n",
      "  * [ ] _cov_13\n",
      "  * [ ] _cov_14\n",
      "  * [ ] _cov_15\n",
      "  * [ ] _cov_16\n",
      "  * [ ] _cov_17\n",
      "  * [ ] _cov_18\n",
      "  * [ ] _cov_19\n",
      "  * [ ] _cov_20\n",
      "  * [ ] _cov_21\n",
      "  * [ ] _cov_22\n",
      "  * [ ] _cov_X\n",
      "  * [ ] _cov_Y\n",
      " Files:\n",
      "  * [X] vcf : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.vcf.bgz\n",
      "  * [X] tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.vcf.bgz.tbi\n",
      "  * [X] cov_1 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.1.tsv.bgz\n",
      "  * [X] cov_1_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.1.tsv.bgz.tbi\n",
      "  * [X] cov_2 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.2.tsv.bgz\n",
      "  * [X] cov_2_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.2.tsv.bgz.tbi\n",
      "  * [X] cov_3 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.3.tsv.bgz\n",
      "  * [X] cov_3_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.3.tsv.bgz.tbi\n",
      "  * [X] cov_4 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.4.tsv.bgz\n",
      "  * [X] cov_4_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.4.tsv.bgz.tbi\n",
      "  * [X] cov_5 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.5.tsv.bgz\n",
      "  * [X] cov_5_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.5.tsv.bgz.tbi\n",
      "  * [X] cov_6 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.6.tsv.bgz\n",
      "  * [X] cov_6_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.6.tsv.bgz.tbi\n",
      "  * [X] cov_7 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.7.tsv.bgz\n",
      "  * [X] cov_7_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.7.tsv.bgz.tbi\n",
      "  * [X] cov_8 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.8.tsv.bgz\n",
      "  * [X] cov_8_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.8.tsv.bgz.tbi\n",
      "  * [X] cov_9 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.9.tsv.bgz\n",
      "  * [X] cov_9_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.9.tsv.bgz.tbi\n",
      "  * [X] cov_10 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.10.tsv.bgz\n",
      "  * [X] cov_10_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.10.tsv.bgz.tbi\n",
      "  * [X] cov_11 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.11.tsv.bgz\n",
      "  * [X] cov_11_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.11.tsv.bgz.tbi\n",
      "  * [X] cov_12 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.12.tsv.bgz\n",
      "  * [X] cov_12_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.12.tsv.bgz.tbi\n",
      "  * [ ] cov_13 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.13.tsv.bgz\n",
      "  * [ ] cov_13_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.13.tsv.bgz.tbi\n",
      "  * [X] cov_14 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.14.tsv.bgz\n",
      "  * [X] cov_14_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.14.tsv.bgz.tbi\n",
      "  * [X] cov_15 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.15.tsv.bgz\n",
      "  * [X] cov_15_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.15.tsv.bgz.tbi\n",
      "  * [X] cov_16 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.16.tsv.bgz\n",
      "  * [X] cov_16_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.16.tsv.bgz.tbi\n",
      "  * [X] cov_17 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.17.tsv.bgz\n",
      "  * [X] cov_17_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.17.tsv.bgz.tbi\n",
      "  * [X] cov_18 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.18.tsv.bgz\n",
      "  * [X] cov_18_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.18.tsv.bgz.tbi\n",
      "  * [X] cov_19 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.19.tsv.bgz\n",
      "  * [X] cov_19_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.19.tsv.bgz.tbi\n",
      "  * [X] cov_20 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.20.tsv.bgz\n",
      "  * [X] cov_20_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.20.tsv.bgz.tbi\n",
      "  * [X] cov_21 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.21.tsv.bgz\n",
      "  * [X] cov_21_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.21.tsv.bgz.tbi\n",
      "  * [X] cov_22 : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.22.tsv.bgz\n",
      "  * [X] cov_22_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.22.tsv.bgz.tbi\n",
      "  * [X] cov_X : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.X.tsv.bgz\n",
      "  * [X] cov_X_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.X.tsv.bgz.tbi\n",
      "  * [ ] cov_Y : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.Y.tsv.bgz\n",
      "  * [ ] cov_Y_tbi : /exports/molepi/tgehrmann/data/gnomad_GRCh37/gnomad.coverage.chr.Y.tsv.bgz.tbi\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gnomad = biu.db.Gnomad() #GnoMAD dataset\n",
    "print(gnomad)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record(CHROM=1, POS=324720, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324728, REF=C, ALT=[A])\n",
      "Record(CHROM=1, POS=324729, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=324737, REF=G, ALT=[C])\n",
      "Record(CHROM=1, POS=324738, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=324740, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=324741, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=324748, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324750, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=324751, REF=G, ALT=[T])\n",
      "Record(CHROM=1, POS=324754, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324768, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324769, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=324771, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324782, REF=G, ALT=[T])\n",
      "Record(CHROM=1, POS=324788, REF=A, ALT=[G])\n",
      "Record(CHROM=1, POS=324789, REF=C, ALT=[A])\n",
      "Record(CHROM=1, POS=324790, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=324792, REF=C, ALT=[T, G])\n",
      "Record(CHROM=1, POS=324795, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324800, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324806, REF=A, ALT=[T])\n",
      "Record(CHROM=1, POS=324812, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324814, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=324815, REF=G, ALT=[T, A])\n",
      "Record(CHROM=1, POS=324822, REF=A, ALT=[T, C, G])\n",
      "Record(CHROM=1, POS=324823, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=324826, REF=A, ALT=[G])\n",
      "Record(CHROM=1, POS=324827, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324829, REF=T, ALT=[G])\n",
      "Record(CHROM=1, POS=324833, REF=T, ALT=[A])\n",
      "Record(CHROM=1, POS=324841, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324842, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324845, REF=A, ALT=[C])\n",
      "Record(CHROM=1, POS=324849, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=324856, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324859, REF=A, ALT=[C, T])\n",
      "Record(CHROM=1, POS=324861, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324876, REF=A, ALT=[G])\n",
      "Record(CHROM=1, POS=324877, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324883, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324885, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=324890, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=324891, REF=G, ALT=[T])\n",
      "Record(CHROM=1, POS=324894, REF=A, ALT=[T, G])\n",
      "Record(CHROM=1, POS=324895, REF=C, ALT=[T, G])\n",
      "Record(CHROM=1, POS=324902, REF=G, ALT=[T])\n",
      "Record(CHROM=1, POS=324904, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324918, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324921, REF=A, ALT=[C])\n",
      "Record(CHROM=1, POS=324922, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=324929, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=324937, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=324945, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324946, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324950, REF=C, ALT=[T, A])\n",
      "Record(CHROM=1, POS=324953, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=324955, REF=G, ALT=[C])\n",
      "Record(CHROM=1, POS=324963, REF=G, ALT=[T])\n",
      "Record(CHROM=1, POS=324964, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=324967, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=324979, REF=A, ALT=[G])\n",
      "Record(CHROM=1, POS=324986, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=324995, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=324997, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325002, REF=G, ALT=[T])\n",
      "Record(CHROM=1, POS=325009, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325010, REF=A, ALT=[G])\n",
      "Record(CHROM=1, POS=325013, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=325015, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=325024, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325027, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325034, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=325041, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=325042, REF=A, ALT=[C, G])\n",
      "Record(CHROM=1, POS=325046, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=325052, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=325059, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325060, REF=A, ALT=[G])\n",
      "Record(CHROM=1, POS=325061, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325062, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=325067, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325069, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=325073, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325075, REF=G, ALT=[C])\n",
      "Record(CHROM=1, POS=325076, REF=G, ALT=[A, T])\n",
      "Record(CHROM=1, POS=325077, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=325081, REF=A, ALT=[G])\n",
      "Record(CHROM=1, POS=325085, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325091, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325094, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=325097, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=325106, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=325107, REF=G, ALT=[C])\n",
      "Record(CHROM=1, POS=325109, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325113, REF=A, ALT=[C])\n",
      "Record(CHROM=1, POS=325119, REF=A, ALT=[C])\n",
      "Record(CHROM=1, POS=325125, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=325127, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=325135, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=325147, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=325151, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325152, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325154, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=325155, REF=C, ALT=[A])\n",
      "Record(CHROM=1, POS=325175, REF=A, ALT=[C])\n",
      "Record(CHROM=1, POS=325178, REF=A, ALT=[C])\n",
      "Record(CHROM=1, POS=325179, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=325188, REF=G, ALT=[T])\n",
      "Record(CHROM=1, POS=325193, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=325196, REF=G, ALT=[C])\n",
      "Record(CHROM=1, POS=325205, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325207, REF=G, ALT=[C])\n",
      "Record(CHROM=1, POS=325224, REF=G, ALT=[T])\n",
      "Record(CHROM=1, POS=325225, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325226, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=325246, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325250, REF=A, ALT=[C])\n",
      "Record(CHROM=1, POS=325251, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=325260, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325297, REF=G, ALT=[A])\n",
      "Record(CHROM=1, POS=325309, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325348, REF=C, ALT=[A])\n",
      "Record(CHROM=1, POS=325352, REF=G, ALT=[T])\n",
      "Record(CHROM=1, POS=325387, REF=A, ALT=[C])\n",
      "Record(CHROM=1, POS=325422, REF=C, ALT=[G])\n",
      "Record(CHROM=1, POS=325476, REF=C, ALT=[T])\n",
      "Record(CHROM=1, POS=325547, REF=A, ALT=[T])\n",
      "Record(CHROM=1, POS=325603, REF=T, ALT=[C])\n",
      "Record(CHROM=1, POS=325635, REF=G, ALT=[T])\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D: VCF Input source is tabixed file.\n",
      "D: VCF Input source is list of Records.\n"
     ]
    }
   ],
   "source": [
    "for record in gnomad.query(1, 324719, 524720, types=['snp']):\n",
    "    print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             id     RR  R     RA  A     AA  O\n",
      "0  1-324950-C-A  39986  0  39986  0  39986  0\n",
      "             id     RR  R  RA  A  AA  O\n",
      "0  1-324950-C-T  39986  0   0  0   1  0\n",
      "             id     RR  R  RA  A  AA  O\n",
      "0  1-324950-C-A  39986  0   1  0   0  0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D: VCF Input source is list of Records.\n",
      "D: VCF Input source is list of Records.\n",
      "D: VCF Input source is list of Records.\n"
     ]
    }
   ],
   "source": [
    "# Alternative allele is reference:\n",
    "#1-324950-C-A  39986  0  39986  0  39986  0\n",
    "print(gnomad.summary(gnomad.query(1, 324950-1, 324950), altPos=[0]))\n",
    "\n",
    "# Alternative allele is first alternative:\n",
    "#1-324950-C-A  39986  0   1  0   0  0\n",
    "print(gnomad.summary(gnomad.query(1, 324950-1, 324950), altPos=[1]))\n",
    "\n",
    "# Alternative allele is second alternative:\n",
    "#1-324950-C-T \t39986 \t0 \t0 \t0 \t1 \t0\n",
    "print(gnomad.summary(gnomad.query(1, 324950-1, 324950), altPos=[2]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record(CHROM=1, POS=324950, REF=C, ALT=[T, A]) 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>RR</th>\n",
       "      <th>R</th>\n",
       "      <th>RA</th>\n",
       "      <th>A</th>\n",
       "      <th>AA</th>\n",
       "      <th>O</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-324950-C-A</td>\n",
       "      <td>39986</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             id     RR  R  RA  A  AA  O\n",
       "0  1-324950-C-A  39986  0   1  0   0  0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var, altPos = gnomad.getVar(1, 324950, 'A')\n",
    "print(var, altPos)\n",
    "gnomad.summary([var], altPos=[altPos])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
