{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using the VEP pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import biu as biu\n",
    "where = '/exports/molepi/tgehrmann/data/'\n",
    "biu.config.settings.setWhere(where)\n",
    "\n",
    "lls = biu.db.LLS()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Produce a list of variants, somehow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D: VCF Input source is tabixed file.\n",
      "D: VCF Input source is list of Records.\n"
     ]
    }
   ],
   "source": [
    "varList = lls.query(1, 10483, 10495)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Execute the VEP pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building DAG of jobs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VEP\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using shell: /bin/bash\n",
      "Provided cores: 1\n",
      "Rules claiming more threads will be scaled down.\n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\toutput\n",
      "\t1\tvep\n",
      "\t2\n",
      "\n",
      "rule vep:\n",
      "    input: /exports/molepi/tgehrmann/pipeline_runs/temporary_input/VEP/digest.cf937c5c6b5e98a66f2843a3a98e7072, /exports/molepi/tgehrmann/pipeline_runs/common/VEP/vep_genomes/homo_sapiens_GRCh37\n",
      "    output: /exports/molepi/tgehrmann/pipeline_runs/VEP/77336778a112a33a1cb7f89291df3513/result.txt\n",
      "    jobid: 1\n",
      "\n",
      "Activating conda environment /exports/molepi/tgehrmann/pipeline_runs/conda/56fa40f0.\n",
      "Finished job 1.\n",
      "1 of 2 steps (50%) done\n",
      "\n",
      "rule output:\n",
      "    input: /exports/molepi/tgehrmann/pipeline_runs/VEP/77336778a112a33a1cb7f89291df3513/result.txt\n",
      "    output: /exports/molepi/tgehrmann/pipeline_runs/VEP/77336778a112a33a1cb7f89291df3513/vep_annotations.tsv\n",
      "    jobid: 0\n",
      "\n",
      "Finished job 0.\n",
      "2 of 2 steps (100%) done\n",
      "Complete log: /home/tgehrmann/repos/BIU/docs/.snakemake/log/2018-06-28T141123.255782.snakemake.log\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VEP Pipeline object\n",
      " Where: /home/tgehrmann/repos/BIU/docs/biu/pipelines/vep/Snakefile\n",
      " Output: /exports/molepi/tgehrmann/pipeline_runs/VEP/77336778a112a33a1cb7f89291df3513\n",
      " Config: /exports/molepi/tgehrmann/pipeline_runs/VEP/77336778a112a33a1cb7f89291df3513/config.json\n",
      " Hash: 77336778a112a33a1cb7f89291df3513\n",
      " Completed: Yes\n",
      "\n"
     ]
    }
   ],
   "source": [
    "vepConfig = {'install_species': 'homo_sapiens', \n",
    "             'install_assembly': 'GRCh37'}\n",
    "\n",
    "vep = biu.pipelines.VEP(varList, config=vepConfig)\n",
    "print(vep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>location</th>\n",
       "      <th>allele</th>\n",
       "      <th>gene</th>\n",
       "      <th>feature</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>consequence</th>\n",
       "      <th>cdna_position</th>\n",
       "      <th>cds_position</th>\n",
       "      <th>protein_position</th>\n",
       "      <th>amino_acids</th>\n",
       "      <th>codons</th>\n",
       "      <th>existing_variation</th>\n",
       "      <th>extra</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-10484-C-T</td>\n",
       "      <td>1:10484</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000423562</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=3879;STRAND=-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1-10484-C-T</td>\n",
       "      <td>1:10484</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000438504</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=3879;STRAND=-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1-10484-C-T</td>\n",
       "      <td>1:10484</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000450305</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=1526;STRAND=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1-10484-C-T</td>\n",
       "      <td>1:10484</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=1385;STRAND=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1-10484-C-T</td>\n",
       "      <td>1:10484</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000488147</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=3920;STRAND=-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1-10484-C-T</td>\n",
       "      <td>1:10484</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000515242</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=1388;STRAND=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1-10484-C-T</td>\n",
       "      <td>1:10484</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000518655</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=1390;STRAND=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1-10484-C-T</td>\n",
       "      <td>1:10484</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000538476</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=3927;STRAND=-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1-10484-C-T</td>\n",
       "      <td>1:10484</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000541675</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=3879;STRAND=-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1-10493-C-A</td>\n",
       "      <td>1:10493</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000423562</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=3870;STRAND=-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1-10493-C-A</td>\n",
       "      <td>1:10493</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000438504</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=3870;STRAND=-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1-10493-C-A</td>\n",
       "      <td>1:10493</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000450305</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=1517;STRAND=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1-10493-C-A</td>\n",
       "      <td>1:10493</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=1376;STRAND=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1-10493-C-A</td>\n",
       "      <td>1:10493</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000488147</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=3911;STRAND=-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1-10493-C-A</td>\n",
       "      <td>1:10493</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000515242</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=1379;STRAND=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1-10493-C-A</td>\n",
       "      <td>1:10493</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000518655</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=1381;STRAND=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1-10493-C-A</td>\n",
       "      <td>1:10493</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000538476</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=3918;STRAND=-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1-10493-C-A</td>\n",
       "      <td>1:10493</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000541675</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=3870;STRAND=-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     variant_id location allele             gene          feature  \\\n",
       "0   1-10484-C-T  1:10484      T  ENSG00000227232  ENST00000423562   \n",
       "1   1-10484-C-T  1:10484      T  ENSG00000227232  ENST00000438504   \n",
       "2   1-10484-C-T  1:10484      T  ENSG00000223972  ENST00000450305   \n",
       "3   1-10484-C-T  1:10484      T  ENSG00000223972  ENST00000456328   \n",
       "4   1-10484-C-T  1:10484      T  ENSG00000227232  ENST00000488147   \n",
       "5   1-10484-C-T  1:10484      T  ENSG00000223972  ENST00000515242   \n",
       "6   1-10484-C-T  1:10484      T  ENSG00000223972  ENST00000518655   \n",
       "7   1-10484-C-T  1:10484      T  ENSG00000227232  ENST00000538476   \n",
       "8   1-10484-C-T  1:10484      T  ENSG00000227232  ENST00000541675   \n",
       "9   1-10493-C-A  1:10493      A  ENSG00000227232  ENST00000423562   \n",
       "10  1-10493-C-A  1:10493      A  ENSG00000227232  ENST00000438504   \n",
       "11  1-10493-C-A  1:10493      A  ENSG00000223972  ENST00000450305   \n",
       "12  1-10493-C-A  1:10493      A  ENSG00000223972  ENST00000456328   \n",
       "13  1-10493-C-A  1:10493      A  ENSG00000227232  ENST00000488147   \n",
       "14  1-10493-C-A  1:10493      A  ENSG00000223972  ENST00000515242   \n",
       "15  1-10493-C-A  1:10493      A  ENSG00000223972  ENST00000518655   \n",
       "16  1-10493-C-A  1:10493      A  ENSG00000227232  ENST00000538476   \n",
       "17  1-10493-C-A  1:10493      A  ENSG00000227232  ENST00000541675   \n",
       "\n",
       "   feature_type              consequence cdna_position cds_position  \\\n",
       "0    Transcript  downstream_gene_variant             -            -   \n",
       "1    Transcript  downstream_gene_variant             -            -   \n",
       "2    Transcript    upstream_gene_variant             -            -   \n",
       "3    Transcript    upstream_gene_variant             -            -   \n",
       "4    Transcript  downstream_gene_variant             -            -   \n",
       "5    Transcript    upstream_gene_variant             -            -   \n",
       "6    Transcript    upstream_gene_variant             -            -   \n",
       "7    Transcript  downstream_gene_variant             -            -   \n",
       "8    Transcript  downstream_gene_variant             -            -   \n",
       "9    Transcript  downstream_gene_variant             -            -   \n",
       "10   Transcript  downstream_gene_variant             -            -   \n",
       "11   Transcript    upstream_gene_variant             -            -   \n",
       "12   Transcript    upstream_gene_variant             -            -   \n",
       "13   Transcript  downstream_gene_variant             -            -   \n",
       "14   Transcript    upstream_gene_variant             -            -   \n",
       "15   Transcript    upstream_gene_variant             -            -   \n",
       "16   Transcript  downstream_gene_variant             -            -   \n",
       "17   Transcript  downstream_gene_variant             -            -   \n",
       "\n",
       "   protein_position amino_acids codons existing_variation  \\\n",
       "0                 -           -      -                  -   \n",
       "1                 -           -      -                  -   \n",
       "2                 -           -      -                  -   \n",
       "3                 -           -      -                  -   \n",
       "4                 -           -      -                  -   \n",
       "5                 -           -      -                  -   \n",
       "6                 -           -      -                  -   \n",
       "7                 -           -      -                  -   \n",
       "8                 -           -      -                  -   \n",
       "9                 -           -      -                  -   \n",
       "10                -           -      -                  -   \n",
       "11                -           -      -                  -   \n",
       "12                -           -      -                  -   \n",
       "13                -           -      -                  -   \n",
       "14                -           -      -                  -   \n",
       "15                -           -      -                  -   \n",
       "16                -           -      -                  -   \n",
       "17                -           -      -                  -   \n",
       "\n",
       "                                      extra  \n",
       "0   IMPACT=MODIFIER;DISTANCE=3879;STRAND=-1  \n",
       "1   IMPACT=MODIFIER;DISTANCE=3879;STRAND=-1  \n",
       "2    IMPACT=MODIFIER;DISTANCE=1526;STRAND=1  \n",
       "3    IMPACT=MODIFIER;DISTANCE=1385;STRAND=1  \n",
       "4   IMPACT=MODIFIER;DISTANCE=3920;STRAND=-1  \n",
       "5    IMPACT=MODIFIER;DISTANCE=1388;STRAND=1  \n",
       "6    IMPACT=MODIFIER;DISTANCE=1390;STRAND=1  \n",
       "7   IMPACT=MODIFIER;DISTANCE=3927;STRAND=-1  \n",
       "8   IMPACT=MODIFIER;DISTANCE=3879;STRAND=-1  \n",
       "9   IMPACT=MODIFIER;DISTANCE=3870;STRAND=-1  \n",
       "10  IMPACT=MODIFIER;DISTANCE=3870;STRAND=-1  \n",
       "11   IMPACT=MODIFIER;DISTANCE=1517;STRAND=1  \n",
       "12   IMPACT=MODIFIER;DISTANCE=1376;STRAND=1  \n",
       "13  IMPACT=MODIFIER;DISTANCE=3911;STRAND=-1  \n",
       "14   IMPACT=MODIFIER;DISTANCE=1379;STRAND=1  \n",
       "15   IMPACT=MODIFIER;DISTANCE=1381;STRAND=1  \n",
       "16  IMPACT=MODIFIER;DISTANCE=3918;STRAND=-1  \n",
       "17  IMPACT=MODIFIER;DISTANCE=3870;STRAND=-1  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep.getAnnotations()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
